### PR TITLE
[cargo-miri] Don't skip `rlib` crates

### LINKED ISSUE #1691: Cannot compile a crate with a `crate-type = ["rlib"]` dependency
See https://github.com/danielhenrymantilla/rust-uninit/pull/4

### Steps to reproduce the problem

Have a crate with a `crate-type = ["rlib"]` dependency (that gets used):

```bash
cd $(mktemp -d)

cargo init --lib --name foo

cargo new --lib bar && echo 'bar = { path = "bar" }' >> Cargo.toml

cat >>bar/Cargo.toml<<\#\ END
[lib]
crate-type = ["rlib"]
# END

echo 'use ::bar;' >> src/lib.rs

cargo miri test
```

The last command fails with:

```text
error: extern location for bar does not exist: /private/var/folders/m_/347zzk7j6nz08tt6z3tnd3z00000gn/T/tmp.c7NSx84d/target/x86_64-apple-darwin/debug/deps/libbar-03d28c0dc71f4bad.rmeta
 --> src/lib.rs:8:7
  |
8 | use ::bar;
  |       ^^^
```

  - When I tried to debug this with a `-vv` flag, the copy-pasted compilation invocation of crate `bar` yielded no compilation artifacts (mainly, no `.rmeta` file).

### Host context

  - As you can see from the error message, this, at least, happens on the `x86_64-apple-darwin` platform; ~~will try to test this on Linux now~~ Same for `x86_64-unknown-linux-gnu`.

  - ```text
    cargo -V
    cargo 1.51.0-nightly (783bc43c6 2021-01-20)
    ```

### PR DESCRIPTION
**Pull request by hyd-dev**
Fixes #1691.

### CODE REVIEW

File: test-cargo-miri/Cargo.lock [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
+version = 3
+
```
Comment (hyd-dev) [inline test-cargo-miri/Cargo.lock:4]: This looks unrelated... Note that I didn't setup the toolchain using `rustup-toolchain-install-master`, but `rustup override set nightly` instead (because don't want to download another toolchain and don't need to build Miri itself), so this may be inconsistent with what `cargo +miri miri` produces. Maybe I should remove this?

File: test-cargo-miri/src/lib.rs [hunk 2]
```diff
@@ -3,5 +3,5 @@
 /// assert!(cargo_miri_test::make_true());
 /// ```
 pub fn make_true() -> bool {
-    true
+    rlib_dep::use_me()
```
Comment (hyd-dev) [inline test-cargo-miri/src/lib.rs:6]: Confirmed that current nightly `cargo-miri` fails here:
```
## Running `cargo miri` tests
A libstd for Miri is now available in `/my/home/directory/.cache/miri/HOST`.
Testing `cargo miri run` (no isolation)...
--- BEGIN stdout ---
--- END stdout ---
--- BEGIN stderr ---
error: extern location for rlib_dep does not exist: /my/home/directory/miri/test-cargo-miri/target/x86_64-unknown-linux-gnu/debug/deps/librlib_dep-095fd0061894c932.rmeta
 --> src/lib.rs:6:5
  |
6 |     rlib_dep::use_me()
  |     ^^^^^^^^

error: aborting due to previous error

error: could not compile `cargo-miri-test`

To learn more, run the command again with --verbose.
--- END stderr ---

TEST FAIL: exit code was 101
```

File: cargo-miri/bin.rs [hunk 3]
```diff
@@ -596,16 +596,13 @@ fn phase_cargo_rustc(args: env::Args) {
...
-            if verbose {
-                eprint!("[cargo-miri rustc] (rlib/cdylib skipped)");
-            }
-            return;
+    if get_arg_flag_value("--crate-type").as_deref() == Some("cdylib") {
+        if verbose {
+            eprint!("[cargo-miri rustc] (cdylib skipped)");
         }
-        _ => {},
+        return;
```
Comment (hyd-dev) [inline cargo-miri/bin.rs:605]: I found that skipping the crate if `get_arg_flag_value("--crate-type")` is something is probably wrong, because Cargo passes *multiple* `--crate-type` to the compiler.

If I add
```rust
issue_1567::use_the_dependency();
```
to `test-cargo-miri/src/lib.rs`, the test will fail:
```
## Running `cargo miri` tests
A libstd for Miri is now available in `/my/home/directory/.cache/miri/HOST`.
Testing `cargo miri run` (no isolation)...
--- BEGIN stdout ---
--- END stdout ---
--- BEGIN stderr ---
error: extern location for issue_1567 does not exist: /my/home/directory/miri/test-cargo-miri/target/x86_64-unknown-linux-gnu/debug/deps/libissue_1567.rmeta
 --> src/lib.rs:6:5
  |
6 |     issue_1567::use_the_dependency();
  |     ^^^^^^^^^^

error: aborting due to previous error

error: could not compile `cargo-miri-test`

To learn more, run the command again with --verbose.
--- END stderr ---

TEST FAIL: exit code was 101
```
I think that is because Cargo passes *both* `--crate-type cdylib` and `--crate-type rlib` to the compiler and only invokes the compiler once. If `cargo-miri` skips the crate here, the `rlib` won't be produced.

Maybe just dropping `--crate-type cdylib` and continuing to compile with only `--crate-type rlib` (or skip the compilation if there's no more `--crate-type`) makes more sense.

File: test-cargo-miri/Cargo.lock (hunk 1 above)
Comment (RalfJung) [reply test-cargo-miri/Cargo.lock:4]: Yeah I think for now it'd be better to just remove the lockfile change; I#d like to wait a bit before updating the checked-in lockfiles to the new format.

File: test-cargo-miri/Cargo.toml [hunk 4]
```diff
@@ -9,6 +9,7 @@ edition = "2018"
 
 [dependencies]
 byteorder = "1.0"
+rlib-dep.path = "rlib-dep"
```
Comment (RalfJung) [inline test-cargo-miri/Cargo.toml:12]: Maybe call this crate `issue_1691` for consistency.
Also please avoid using two different syntaxes for path dependencies here.

File: cargo-miri/bin.rs (hunk 3 above)
Comment (RalfJung) [reply cargo-miri/bin.rs:605]: > (EDIT: Hit #1705 when experimenting with that... Trying to fix that together...)

So I suggest we tackle that in a future MR. Good catch though, this should be added to the test suite then (together with a `crate-type = ["lib", "staticlib", "cdylib"]` crate that triggers [this](https://github.com/rust-lang/miri/issues/1705#issuecomment-778651455) behavior).

**Comment by RalfJung** [comment]
Thanks for the PR. :)

**Comment by RalfJung** [comment]
Looking good, thanks for your help in working out the kinks of cargo-miri!
@bors r+
